crates/ty_python_semantic/src/types.rs

```diff
@@ -4762,7 +4764,40 @@ impl<'db> Type<'db> {
                 // TODO: Use an opt-in rule for a bare `Callable`
                 KnownInstanceType::Callable => Ok(Type::Callable(CallableType::unknown(db))),
 
-                KnownInstanceType::TypingSelf => Ok(todo_type!("Support for `typing.Self`")),
+                KnownInstanceType::TypingSelf => {
+                    let index = semantic_index(db, scope_id.file(db));
+                    let Some(class_ty) = enclosing_class_symbol(db, index, scope_id) else {
+                        return Err(InvalidTypeExpressionError {
+                            fallback_type: Type::unknown(),
+                            invalid_expressions: smallvec::smallvec![
+                                InvalidTypeExpression::InvalidType(*self)
+                            ],
+                        });
+                    };
+                    let Some(TypeDefinition::Class(class_def)) = class_ty.definition(db) else {
+                        debug_assert!(
+                            false,
+                            "enclosing_class_symbol must return a type with class definition"
+                        );
+                        return Ok(Type::unknown());
+                    };
+                    let Some(instance) = class_ty.to_instance(db) else {
+                        debug_assert!(
+                            false,
+                            "enclosing_class_symbol must return type that can be instantiated"
+                        );
```

Comment:
I am not 100% sure that these debug-asserts will always remain true, in the face of e.g. class decorators. But we also haven't fully decided yet how to handle class decorators. I think it's fine to have these here for now; they'll at least force us to consider the impact here if we ever do violate these invariants.

Reply:
I think another option here would be to split the `enclosing_class_symbol` function in half, and turn the first half of it into `enclosing_class_definition` instead, returning a `Definition`, and use a separate function to get the inferred type for that definition. Then you'd have the definition in hand already, and wouldn't have to pull it back out of the type.